pkg/pbq/pbq.go

```diff
@@ -0,0 +1,101 @@
+package pbq
+
+import (
+	"context"
+	"errors"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/pbq/store"
+	"go.uber.org/zap"
+)
+
+var COBErr error = errors.New("error while writing to pbq, pbq is closed")
+
+type PBQ struct {
+	store       store.Store
+	output      chan *isb.Message
+	cob         bool // cob to avoid panic in case writes happen after close of book
+	partitionID string
+	options     *Options
+	manager     *Manager
+	log         *zap.SugaredLogger
+}
+
+// Write writes message to pbq and persistent store
+// We don't need a context here as this is invoked for every message.
+func (p *PBQ) Write(ctx context.Context, message *isb.Message) (writeErr error) {
+	// if cob we should return
+	if p.cob {
+		p.log.Errorw("failed to write message to pbq, pbq is closed", zap.Any("partitionID", p.partitionID), zap.Any("header", message.Header))
+		writeErr = COBErr
+		return
+	}
+	// we need context to get out of blocking write
+	select {
+	case p.output <- message:
+		writeErr = p.store.Write(message)
+		return
+	case <-ctx.Done():
+		// closing the output channel will not cause panic, since its inside select case
+		close(p.output)
+		writeErr = p.store.Close()
```

Comment:
```suggestion
		writeErr = ctx.Err()
```
`ctx.Done` implicitly means that the write hasn't succeeded.